File: packages/sdk-ts/src/resources/campaigns.ts
import type { Http, RequestOptions } from '../core/http.js';
import type { Page } from '../core/page.js';
import type { Campaign, CreateCampaign, PaginationQuery } from '../types.js';

const enc = encodeURIComponent;

/** Campaigns: one agent calling a list of leads. */
export class Campaigns {
    constructor(private readonly http: Http) {}

    list(query: PaginationQuery = {}, opts?: RequestOptions): Promise<Page<Campaign>> {
        return this.http.list<Campaign>('/v1/campaigns', query, opts);
    }

    /** Creating does not dial anyone; call `start()` when ready to spend. */
    create(body: CreateCampaign, opts?: RequestOptions): Promise<Campaign> {
        return this.http.request<Campaign>('POST', '/v1/campaigns', { ...opts, body });
    }

    retrieve(id: string, opts?: RequestOptions): Promise<Campaign> {
        return this.http.request<Campaign>('GET', `/v1/campaigns/${enc(id)}`, opts);
    }

    delete(id: string, opts?: RequestOptions): Promise<void> {
        return this.http.request<void>('DELETE', `/v1/campaigns/${enc(id)}`, opts);
    }

    /** Begin calling every lead. Spends credit; refused with `insufficient_credits` when the balance cannot cover the run. */
    start(id: string, opts?: RequestOptions): Promise<Campaign> {
        return this.http.request<Campaign>('POST', `/v1/campaigns/${enc(id)}/start`, opts);
    }

    /** Stop placing new calls and hang up any in flight. */
    pause(id: string, opts?: RequestOptions): Promise<Campaign> {
        return this.http.request<Campaign>('POST', `/v1/campaigns/${enc(id)}/pause`, opts);
    }
}

File: packages/sdk-dart/lib/src/resources/campaigns.dart
import '../models/campaign.dart';
import '../page.dart';
import '../transport.dart';

/// Campaigns: one agent calling a list of leads.
class Campaigns {
  /// Binds the resource to a transport.
  const Campaigns(this._transport);

  final Transport _transport;

  /// Lists campaigns.
  Future<Page<Campaign>> list(
          {int? limit, String? startingAfter, String? endingBefore}) =>
      _transport.list('/v1/campaigns', Campaign.fromJson, query: {
        'limit': limit,
        'starting_after': startingAfter,
        'ending_before': endingBefore,
      });

  /// Creates a campaign. Creating does not dial anyone; call [start].
  Future<Campaign> create(CreateCampaign body) => _transport
      .request('POST', '/v1/campaigns', Campaign.fromJson, body: body.toJson());

  /// Fetches one campaign.
  Future<Campaign> retrieve(String id) => _transport.request(
      'GET', '/v1/campaigns/${Uri.encodeComponent(id)}', Campaign.fromJson);

  /// Deletes a campaign. Pause it first if it is running.
  Future<void> delete(String id) => _transport.requestVoid(
      'DELETE', '/v1/campaigns/${Uri.encodeComponent(id)}');

  /// Begins calling every lead. Spends credit; refused with
  /// `insufficient_credits` when the balance cannot cover the run.
  Future<Campaign> start(String id) => _transport.request('POST',
      '/v1/campaigns/${Uri.encodeComponent(id)}/start', Campaign.fromJson);

  /// Stops placing new calls and hangs up any in flight.
  Future<Campaign> pause(String id) => _transport.request('POST',
      '/v1/campaigns/${Uri.encodeComponent(id)}/pause', Campaign.fromJson);
}
